go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -0,0 +1,1768 @@
+		// 4          |    -                   |    e1                     |    -
+
+		// Case 1 handles updates. Case 2 and 3 makes sure we don't add until we
+		// successfully delete. Case 4 just shows an example of what would
+		// happen if we don't have anything to add
+		ipsToAssign = ipsToAssign.Intersection(ipsToRemove)
+	}
+
+	if !util.PlatformTypeIsEgressIPCloudProvider() {
+		if len(statusToRemove) > 0 {
+			// Delete the statusToRemove from the allocator cache. If we don't
+			// do this we will occupy assignment positions for the ipsToAssign,
+			// even though statusToRemove will be removed afterwards
+			eIPC.deleteAllocatorEgressIPAssignments(statusToRemove)
+		}
+		if len(ipsToAssign) > 0 {
+			statusToAdd = eIPC.assignEgressIPs(name, ipsToAssign.UnsortedList())
+			statusToKeep = append(statusToKeep, statusToAdd...)
+		}
+		// Add all assignments which are to be kept to the allocator cache,
+		// allowing us to track all assignments which have been performed and
+		// avoid incorrect future assignments due to a de-synchronized cache.
+		eIPC.addAllocatorEgressIPAssignments(name, statusToKeep)
+		// Update the object only on an ADD/UPDATE. If we are processing a
+		// DELETE, new will be nil and we should not update the object.
+		if len(statusToAdd) > 0 || (len(statusToRemove) > 0 && new != nil) {
+			if err := eIPC.patchReplaceEgressIPStatus(name, statusToKeep); err != nil {
+				return err
+			}
+		}
+	} else {
+		// Even when running on a public cloud, we must make sure that we unwire EgressIP
+		// configuration from OVN *before* we instruct the CloudNetworkConfigController
+		// to remove the CloudPrivateIPConfig object from the cloud.
+		// CloudPrivateIPConfig objects can be in the "Deleting" state for a long time,
+		// waiting for the underlying cloud to finish its action and to report success of the
+		// unattach operation. Some clouds such as Azure will remove the IP address nearly
+		// immediately, but then they will take a long time (seconds to minutes) to actually report
+		// success of the removal operation.
+		if len(statusToRemove) > 0 {
+			// Delete all assignments that are to be removed from the allocator
+			// cache. If we don't do this we will occupy assignment positions for
+			// the ipsToAdd, even though statusToRemove will be removed afterwards
+			eIPC.deleteAllocatorEgressIPAssignments(statusToRemove)
+			// Before updating the cloud private IP object, we need to remove the OVN configuration
+			// for these invalid statuses so that traffic is not blackholed to non-existing setup in the
+			// cloud. Thus we patch the egressIP status with the valid set of statuses which will
+			// trigger an event for the ovnkube-master to take action upon.
+			// Note that once we figure out the statusToAdd parts below we will trigger an
+			// update to cloudPrivateIP object which will trigger another patch for the eIP object.
+			if err := eIPC.patchReplaceEgressIPStatus(name, statusToKeep); err != nil {
```

Comment:
@kyrtapz : It turned out to be quiet simple..... though I had to double check what each variable was doing.. does this make sense to you?
from my understanding:

1. `status` has the fresh set of statuses from egressIP spec
2. `statusToKeep` at this stage has valid statuses -> `status` minus `invalidStatus`
3. `statusToRemove` at this stage has invalid statuses -> `status` minus `validStatus`

At this point we don't need to consider ipsToRemove and ipsToAdd I think? because that will be done after `executeCloudPrivateIPConfigChange` ? Here we only care about ensuring we are removing the invalid ones... Based on that the logic I took was to update the statusToKeep. WDYT?

Reply:
Makes sense to me.
In the previous code this is how we handled statusToRemove for cloud:
```
if len(statusToRemove) > 0 {
	// Delete all assignments that are to be removed from the allocator
	// cache. If we don't do this we will occupy assignment positions for
	// the ipsToAdd, even though statusToRemove will be removed afterwards
	oc.deleteAllocatorEgressIPAssignments(statusToRemove)
	if err := oc.deleteEgressIPAssignments(name, statusToRemove); err != nil {
		return err
	}
}
...
if err := oc.executeCloudPrivateIPConfigChange(name, statusToAdd, statusToRemove); err != nil {
	return err
}
```
To achieve something similar we need to update the EIP status so it doesn't include statusToRemove.